### PR TITLE
feat: request service

### PR DESCRIPTION
**Pull request by lengau**
This service wraps Requests and adds some convenience functions methods. It's used by the remote build service, but can be used any time one might need requests.

- [x] Have you followed the guidelines for contributing?
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [x] Have you successfully run `tox`?

-----

### CODE REVIEW

File: tests/integration/services/test_request.py
Comment (mr-cal) [inline tests/integration/services/test_request.py:1]: Do you think it's worthwhile to test downloading multiple files?

File: tests/unit/services/test_request.py [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+        ]
+    )
+
+
+@responses.activate
+@pytest.mark.parametrize(
+    "downloads",
+    [
+        {
+            "http://example/file": b"abc",
```
Comment (mr-cal) [inline tests/unit/services/test_request.py:86]: What about testing an empty file?

**Review by mr-cal — APPROVED** [review]
I tried it myself and it looks good!

File: tests/unit/services/test_request.py (hunk 1 above)
Comment (lengau) [reply tests/unit/services/test_request.py:87]: Done! But I've only tested it here because it requires more logic in the test (empty files don't get an `advance` call).

File: tests/integration/services/test_request.py
Comment (lengau) [reply tests/integration/services/test_request.py:1]: Good point. I've added a test that grabs all the (current) `craft-application` releases from GitHub and compares them to checksums I generated myself.

File: craft_application/services/request.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+                yield int(download.headers.get("Content-Length", -1))
+                for chunk in download.iter_content(None):
+                    file.write(chunk)
+                    yield len(chunk)
+
+    def download_with_progress(self, url: str, dest: pathlib.Path) -> pathlib.Path:
+        """Download a single file with a progress bar."""
+        return self.download_files_with_progress({url: dest})[url]
+
+    def download_files_with_progress(
```
Comment (tigarmo) [inline craft_application/services/request.py:76]: very nice!

File: tests/integration/services/test_request.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+import pytest
+import pytest_check
+
+
+@pytest.mark.parametrize(
+    ("url", "checksum", "size"),
+    [
+        pytest.param(
+            # A real build log.
+            "https://launchpad.net/~charmcraft-team/charmcraft/+snap/charmcraft-edge/+build/2377699/+files/buildlog_snap_ubuntu_jammy_armhf_charmcraft-edge_BUILDING.txt.gz",
```
Comment (tigarmo) [inline tests/integration/services/test_request.py:29]: Do we know if this file is going to stay there forever?

**Review by tigarmo — APPROVED** [review]
looks great! I left a comment about the reliability of the remote files used in integration tests but it's minor(ish) and, if we decide to change it, we can do it later

File: tests/integration/services/test_request.py (hunk 3 above)
Comment (lengau) [reply tests/integration/services/test_request.py:29]: As far as I can tell public build logs are available indefinitely, but it's possible that if we delete that recipe from launchpad it'll disappear.
